added support for setting a RandomState other than np.random

mheilman opened the pull request:
This is intended to address #6.

ebenolson commented:
I think it would be better to assign `rng = np.random` in `lasagne/__init__.py`, and omit the getter/setter.

dnouri commented:
> I think it would be better to assign rng = np.random in 
> lasagne/__init__.py, and omit the getter/setter.
> 

The problem with this is that code that does `from lasagne import rng` 
before other code that sets `lasagne.rng` to a custom value will still 
refer to the old `rng`.  It's error-prone.

f0k commented on lasagne/layers/noise.py:
```diff
@@ -54,7 +54,7 @@ class DropoutLayer(Layer):
     def __init__(self, incoming, p=0.5, rescale=True, **kwargs):
         super(DropoutLayer, self).__init__(incoming, **kwargs)
-        self._srng = RandomStreams(np.random.randint(1, 2147462579))
+        self._srng = RandomStreams(get_rng().randint(1, 2147462579))
```
Should we wrap this as `get_srng()`, or `create_srng()`?

f0k commented:
> code that does `from lasagne import rng`

Good catch! So we need to go with setters/getters.

In theory, we could even use different RNGs for weights and noise layers, via `get_rng('init')` and `get_rng('noise')`, and something like:
```python
rngs = {'': np.random}
def set_rng(rng, name=''):
    rngs[name] = rng
def get_rng(name=''):
    rngs.get(name, default=rngs[''])
```

Would that be helpful? It would be easier than explicitly passing RNGs for different purposes (e.g., as part of instantiating an initializer or layer), but almost as flexible.

dnouri commented:
> we could even use different RNGs for weights and noise layers

Your code looks good, though I can't think of a case where I'd want to do have different generators for init and noise.

ebenolson commented:
> The problem with this is that code that does `from lasagne import rng`
before other code that sets `lasagne.rng` to a custom value will still
refer to the old `rng`.  It's error-prone.

Good point, but I think it's more pythonic to just have an `rng.py` that consists of `from numpy import random as rng`, and refer to `rng.rng` where needed.

dnouri commented:
> Good point, but I think it's more pythonic to just have an rng.py that consists of from numpy import random as rng, and refer to rng.rng where needed.

I personally prefer from-imports that import module globals, so it would trip *me* up.  I don't think there's any official recommendation for one over the other.

f0k commented on lasagne/__init__.py:
```diff
@@ -14,6 +14,7 @@
 else:
     del theano
 
+
```
why?

f0k commented on lasagne/init.py:
```diff
@@ -4,13 +4,14 @@
 >>> from lasagne.layers import DenseLayer
->>> from lasagne.init import Constant, Glorot
+>>> from lasagne.init import Constant
 >>> l1 = DenseLayer((100,20), num_units=50, W=GlorotUniform(), b=Constant(0.0))
```
It should be `Constant, GlorotUniform`. It still works without the import statement because it's the current module, but for people to be able to copy/paste it, we should include it.

f0k commented:
> Your code looks good, though I can't think of a case where I'd want to do have different generators for init and noise.

Me neither, at least none that is terribly realistic. It could be interesting when adding training loop support to Lasagne, though, and it was mostly meant as an additional argument for the getter/setter approach (because it allows us to extend that approach to support multiple rngs later, without breaking compatibility).

> I personally prefer from-imports that import module globals, so it would trip *me* up.

That's enough of a reason then!

I'm fine with merging, except that I'd like to hear other opinions on providing a `get_srng()` or `make_srng()` or `create_srng()` convenience function. Also I haven't checked test coverage yet, is it still 100%? coveralls is down for the moment.

ebenolson commented:
> I personally prefer from-imports that import module globals, so it would trip me up. I don't think there's any official recommendation for one over the other.

https://docs.python.org/2/howto/doanddont.html#from-module-import-name1-name2

dnouri commented:
> https://docs.python.org/2/howto/doanddont.html#from-module-import-name1-name2

Interesting.  Had no idea this is recommended against anywhere.  I guess I'll continue doing it though; saves me from typing so much.

mheilman commented:
@ebenolson, thanks for the suggestion on making it more pythonic.  I was having trouble working with the issue of importing module globals (e.g., how `from rng import rng` and then `rng = RandomState(42)` doesn't work), but I think it should work now.  I added a note in the docstring for `random.py` so that users hopefully don't make that mistake.

I also changed `rng.py` to `random.py` because I figured `lasagne.random.rng = rng1` looks better than `lasagne.rng.rng = rng1`.

dnouri commented:
Let's go back to using getter and setter here.  Because we might be able to decide what's Pythonic or not *for us*.  But how users will import and use that global is another story.  We should avoid this non-obvious pitfall; it's also a hard one to figure out.

+1 to renaming the module to `random`.

mheilman commented:
Sure, I can bring back the getter/setter.  I don't have a strong preference for either approach.

I'll wait a bit to see if anybody feels strongly either way.

dnouri commented:
> Incidentally, what I wanted to do at first was have a module-level property...

Interesting.  [It turns out](http://stackoverflow.com/questions/880530/can-python-modules-have-properties-the-same-way-that-objects-can/880751#880751) you can inherit from ModuleType and put an instance of your Module class (with properties!) into `sys.modules`.

But yeah, let's not go there.  :-)

ebenolson commented:
> Let's go back to using getter and setter here. Because we might be able to decide what's Pythonic or not for us. But how users will import and use that global is another story. We should avoid this non-obvious pitfall; it's also a hard one to figure out.

Ideally being pythonic is following an idiom that is widely known/accepted, not a personal judgment.
[I feel pretty certain what the right way to do this is](https://docs.python.org/2/faq/programming.html#how-do-i-share-global-variables-across-modules), but  it's really a very minor issue, so I'll leave it here.

dnouri commented:
@ebenolson This stuff is no way "widely known" and too easy to get wrong.  If another person can share their opinion, we can lay this issue to rest.

f0k commented:
I'm still on the getter/setter side a) to avoid the pitfall and b) to allow extending it to support multiple RNGs. I've also googled for module-level properties some days ago and found a similar stackoverflow thread as Daniel, but that's not a viable alternative :)
I'm fine with renaming the module to `lasagne.random` as well. And I'd still like to hear other's opinions on adding a `get_srng()`, `create_srng()` or `make_srng()` function performing what's currently done in the constructor of the two noise layers.

Hint: The easiest might be a) to `git reset --hard` to the "remove extra newline" commit, `git cherry-pick` the "added GlorotUniform ..." commit and then rename the module, or b) to `git revert` the "more pythonic" commit and then rename the module. Before merging, the commits should be squashed anyway.

benanne commented:
getter/setter seems like the safest bet, `lasagne.random` sounds good. But take my opinion with a grain of salt, I haven't had time to review the entire discussion.

mheilman commented:
I reverted the "more pythonic" commit, renamed `rng.py` to `random.py`, and rebased it into 1 commit.

I don't think methods like `get_srng` are necessary, but I don't mind adding them if people would prefer that.  Or, somebody else could add them pretty easily later in another PR if we change our minds.

f0k commented on lasagne/random.py:
```diff
@@ -0,0 +1,26 @@
+def get_rng():
+    """Get the package-level random number generator.
+    :return: a numpy.random or a numpy.random.RandomState object
+    """
```
We're using numpydoc throughout the package, see [the numpydoc howto](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt) for a detailed treatment or any file in Lasagne for examples. Here, I'd suggest:
```
Get the package-level random number generator for Lasagne.

Returns
-------
:class:`numpy.random.RandomState` instance
    The :class:`numpy.random.RandomState` instance passed to the most
    recent call of :func:`set_rng`, or ``np.random`` if :func:`set_rng` has
    never been called.
```

f0k commented:
> I reverted the "more pythonic" commit, renamed `rng.py` to `random.py`, and rebased it into 1 commit.

Nice, thanks! I'm fine with postponing `get_srng()` until we have more use cases for it. The last request I'd like to make is changing the docstrings to numpydoc syntax, and including it in the documentation (this just needs a `docs/modules/random.rst` file that you can copy/paste and adapt from `utils.rst`, and a TOC entry in `docs/index.rst`, I guess I'd put it before `utils`). Oh yes, and including it in `__init__.py` (that one doesn't seem to have any particular order for some reason).

mheilman commented:
Ah, right, sorry about the improperly formatted docstrings, etc.  I'm not very familiar with numpydoc, so let me know if I need to correct anything.

f0k commented on lasagne/random.py:
```diff
@@ -0,0 +1,36 @@
+    :class:`numpy.random.RandomState` instance
+        The :class:`numpy.random.RandomState` instance passed to the most
+        recent call of :func:`set_rng`, or `numpy.random` if :func:`set_rng` has
+        never been called.
```
`has` needs to be moved into the next line, otherwise it's too long (that's what Travis is complaining above).

While we're at it, `numpy.random` needs double apostrophes as in [my suggestion](https://github.com/Lasagne/Lasagne/pull/343#discussion_r35886511). Same for the module docstring above. The others are correct! (It takes a while to get used to Sphinx' ReST markup.)

mheilman commented:
Sorry again about the documentation issues.   I was confused by this sentence saying module names should be in single ticks: ```Use *italics*, **bold** and ``monospace`` if needed in any explanations (but not for variable names and doctest code or multi-line code). Variable, module, function, and class names should be written between single back-ticks (`numpy`)``` (https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt), but I guess that only works for classes and modules within the package (or with intersphinx?).

I built the docs (as I should have done before!), and I think it looks OK now.  Please let me know if you notice anything else.

By the way, when I ran sphinx, I needed to pip install `sphinx_rtd_theme`.  Would it be worth adding that to `requirements-dev.txt`?

f0k commented:
> I guess that only works for classes and modules within the package (or with intersphinx?).

Yes, I think that's only when using the `:mod:`, `:class:` or `:func:` prefixes. When not using a prefix, it will be treated as a variable name (so that's what should be used for referencing function arguments in a docstring). intersphinx would be useful to actually make `numpy.random` link to something, but we currently don't use it.

> By the way, when I ran sphinx, I needed to pip install `sphinx_rtd_theme`. Would it be worth adding that to `requirements-dev.txt`?

Yes... I removed it in #257 because it's included with Sphinx, but it's actually *not* included with the version of Sphinx we've got in `requirements-dev.txt` (my bad), so we should add it back. This should be a separate PR, though.

> I think it looks OK now.

Yes, thanks a lot! I'm fine with merging this :+1: Let's wait a bit if anybody else wants to comment.

f0k commented:
> This should be a separate PR, though.

Or maybe I'll just add that to #335 because it touches the configuration a bit.

f0k commented:
Merged, thanks again for your contribution!
